### ISSUE TITLE
defer crashes in closure for os.walkDir()

### ISSUE THREAD

**Issue by mjfh** [nim-lang/Nim#4966]
The following code crashes on line 6 when wrapping os.walkDir() into a closure

```
1  import os-
2
3  proc walk(obj: string): iterator(): string =
4    result = iterator(): string {.closure.} =
5      for k,p in walkDir(obj):
6        yield p
7
8  var g = walk(".")
9  while not finished(g):
10   echo "---> ", g()
```

The reason for crashing seems to come from a 'defer closedir()' directive, see line 974 in lib/pure/os.nim

```
...
972    var d = opendir(dir)
973    if d != nil:
974      defer: discard closedir(d)
975      while true:
976        var x = readdir(d)
977        if x == nil: break
978        var y = $x.d_name
...
```

Commenting out line 974, the os.walkDir() function works (does not crash). Replacing the defer directive by

```
974      defer: echo "===>", repr(d); discard closedir(d)
```

shows that the defer command is invoked early before all yields are exhausted.

Maybe this is correct and there is some caching in the background I am not aware of which holds all the rest of the iterator results already. But then again the cache was flushed too early before all values had been yielded.

Please excuse my double posting on the forum already when I was not sure whether I did something wrong (I am quite new to NIM - thanks for providing it).

Jordan

**Comment by dom96** [comment]
huh, this seems like a compiler bug to me. @Araq please take a look.

**Comment by Araq** [comment]
This is just a 'yield' within a 'try' which is not supported.

**Comment by dom96** [comment]
@narimiran Please explain why you've closed an issue when closing it. It might be obvious to you, but not to others.

My guess is that this was closed because ``yield`` in a try statement is now valid, so the code in this issue works now. It would be great if you could confirm that.

**Comment by narimiran** [comment]
Commit 4644484 closed #5321 (by having a test for it), which was marked as duplicate of this one. I've run both issues and both work now.
